### PR TITLE
add OAuth authentication

### PR DESCRIPTION
**Pull request by kurddt**
### Your checklist for this pull request

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Did you provide a test-case? That demonstrates feature works or fixes the issue.

I tried to revive #69 by re-basing on master and adapting the code to the changes that was made since especially #101 .

The functionality is working, but the normal user/password is broken. I don't really have any idea how to make both work so any suggestion would be greatly appreciated

### CODE REVIEW

**Comment by casz** [comment]
Thanks for picking up the effort :)

**Comment by kurddt** [comment]
@casz Any idea on whom could help me / give me some pointer on how to support both OAuth and normal credential ?

**Comment by kurddt** [comment]
I made a test case for the OAuth authenticator but it requires an access to bitbucket.org and to generate a OAuth key/secret

**Comment by casz** [comment]
I'd appreciate if you could avoid force pushing 😓

**Comment by kurddt** [comment]
Sure. Any idea why the integration failed on jenkins ?

**Comment by casz** [comment]
@kurddt I'd guess ci.jenkins.io is having a bad day :)

https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fbitbucket-branch-source-plugin/detail/PR-262/9/tests

**Comment by kurddt** [comment]
@casz  Is there any way to re-start it ?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        return !isEMail && validKeyLenght && validSecretLenght;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String describe() {
+        // TODO Auto-generated method stub
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java]: Hmm does this TODO need to be there?
```suggestion
```

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    private static int secretLenght = 32;
+
+    private static final long serialVersionUID = 6458784517693211197L;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean matches(Credentials item) {
+        // TODO Auto-generated method stub
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java]:
```suggestion
```

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 3]
```diff
@@ -245,6 +245,12 @@ public String getRepositoryUri(@NonNull BitbucketRepositoryType type,
             case GIT:
                 switch (protocol) {
                     case HTTP:
+                        if(authenticator != null ) {
+                            String username = authenticator.getUserUri();
+                            if (!username.isEmpty()) {
+                                return "https://" + username + "@bitbucket.org/" + owner + "/" + repository + ".git";
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: not sure that would work as username would return
```
    public String getUserUri() {
        return "x-token-auth:{" + token.getToken() + "}";
    }
```

Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: I guess it does according to the API: https://confluence.atlassian.com/bitbucket/oauth-on-bitbucket-cloud-238027431.html

Comment (kurddt) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Yes. I introduced getUserUri() member method in the authenticator in order to use the retrieved token according to the documentation you linked

File: pom.xml [hunk 4]
```diff
@@ -173,6 +173,16 @@
...
+   </dependency>
+   <dependency>
+      <groupId>org.scribe</groupId>
+      <artifactId>scribe</artifactId>
+      <version>1.3.3</version>
+    </dependency>
+    <dependency>
+      <groupId>commons-codec</groupId>
+      <artifactId>commons-codec</artifactId>
+      <version>1.10</version>
```
Comment (casz) [inline pom.xml]: I believe this dependency should be added to dependency management instead

**Comment by casz** [comment]
Problem atm is both Oauth and username password uses the same base class `StandardUsernamePasswordCredentials`

**Comment by casz** [comment]
I would love some documentation on how to set this up.
Cause I will need the docs to test this 😓

**Comment by kurddt** [comment]
I could write a small documentation, about the usage of this feature just tell me in which format and/or where you want want it. 

Concerning, the base class. Even if they use the same base class, the class can detect if the credentials match a API key and secret. The other solution would be to introduce an other type of credentials in Jenkins but I had no idea how to do so.

**Comment by casz** [comment]
We have section for that on the user guide: https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/master/docs/USER_GUIDE.adoc

File: docs/USER_GUIDE.adoc [hunk 5]
```diff
@@ -98,8 +98,17 @@ two credentials to configure:
...
 image::images/screenshot-3.png[scaledwidth=90%]
+
+=== OAuth credentials
+
+Bitbucket plugin can make use of OAuth credentials instead of the standard username/password.
+
+First create a new OAuth consumer as instructed in https://confluence.atlassian.com/bitbucket/oauth-on-bitbucket-cloud-238027431.html[Bitbucket OAuth Documentation].
+Don't forget to check _This is a private consumer_ and at least allow read access to the repositories and Pull requests. If you want the Bitbucket to install the Webhooks also allow the read and write access of the Webhooks
+
+Then create new _Username with password credentials_, enter the Bitbucket OAuth consumer key in _Username_ field and the Bitbucket OAuth consumer secret in _Password_ field
```
Comment (casz) [inline docs/USER_GUIDE.adoc]: Perhaps give an example and screenshot 🙏

**Comment by casz** [comment]
link to see current commit author: https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/262/commits/13ca375e745844106dba199de36ac202b9959d9f.patch 

@kurddt 
You might want to set up an email alias on your GitHub account 🥇
https://help.github.com/articles/adding-an-email-address-to-your-github-account/

Or look into fixing your git user.email config 😅  
https://help.github.com/articles/setting-your-commit-email-address-in-git
then you would need to amend your commits: https://stackoverflow.com/a/3042512
after changing author

to get proper credit for your commits 🏆

**Comment by kurddt** [comment]
@casz Thanks, I made a new clone of the repo but forgot to change the local git config :sweat_smile:

**Comment by cblakkan** [comment]
It would be awesome to have this functionality supported, did this go stale?

**Comment by jetersen** [comment]
hopefully #299 should remove the blocker

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import hudson.util.Secret;
+import org.apache.http.HttpRequest;
+import org.scribe.model.OAuthConfig;
+import org.scribe.model.OAuthConstants;
+import org.scribe.model.Token;
+
+public class BitbucketOAuthAuthenticator extends BitbucketAuthenticator {
+
+    private Token token;
```
Comment (jetersen) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthAuthenticator.java:13]: Perhaps a good idea to cache the token, I assume it has a validity time 🤔

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 7]
```diff
@@ -245,6 +245,12 @@ public String getRepositoryUri(@NonNull BitbucketRepositoryType type,
             case GIT:
                 switch (protocol) {
                     case HTTP:
+                        if(authenticator != null ) {
```
Comment (jetersen) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java:248]:
```suggestion
                        if (authenticator != null) {
```

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthAuthenticator.java (hunk 6 above)
Comment (jetersen) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthAuthenticator.java:13]: I think we can move forward without caching.

**Pull request merged**